lib/ukfs: Introduce node driver template

andreittr opened the pull request:
### Description of changes

This change introduces the ukfs node driver template, a header with common boilerplate code for implementing the live ref driver API on top of driver code that does little to no runtime bookkeeping internally.

In addition, included are two trivial fixes for the live reference template.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

N/A

andreittr committed:
lib/ukfs: Fix live template type assumption

Previously the live template would assume the nodetype to be of pointer
type, assigning NULL to a such-typed variable.
This change fixes this oversight, allowing for truly generic nodetypes.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/ukfs: Fix handling of lookup returns

Previously the live template lookup relied on asserts to prevent it from
passing up an illegal return value received from the driver lookup.
This change fixes that logic, ensuring that any illegal value
immediately causes a crash, as that indicates a logically faulty driver.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/ukfs: Introduce node driver template

This change introduces the ukfs node driver template, a header with
common boilerplate code for implementing the live ref driver API on top
of driver code that does little to no runtime bookkeeping internally.

Checkpatch-Ignore: SPACING
Checkpatch-Ignore: FUNCTION_ARGUMENTS
Checkpatch-Ignore: OPEN_BRACE
Checkpatch-Ignore: COMPLEX_MACRO
Checkpatch-Ignore: MACRO_ARG_REUSE
Checkpatch-Ignore: MACRO_WITH_FLOW_CONTROL
Checkpatch-Ignore: MULTISTATEMENT_MACRO_USE_DO_WHILE
Signed-off-by: Andrei Tatar <ttr@unikraft.io>

mogasergiu commented on lib/ukfs/include/uk/fs/template/node.h:
```diff
@@ -0,0 +1,1977 @@
+
+/* fs readlink */
+
+#define UK_FS_TMPL_NODE_OP_LIVE_FS_READLINK(drv) dr##_LIVE_FS_READLINK
```
`drv##_LIVE_FS_READLINK`*? (missing "v") :sweat_smile:

andreittr replied:
Whoops :eyes: thanks, that typo would've taken quite a bit to find otherwise :smile:

mogasergiu commented on lib/ukfs/include/uk/fs/template/node.h:
```diff
@@ -0,0 +1,1977 @@
+ * filesystem object until explicitly closed or forgotten.
+ * A node may be "unopened" -- can only perform lookup() -- or "open" -- can
+ * perform all operations.
+ * - vopen() and lookup() return unopened nodes
```
but `vopen` seems to call `_RSTATE_OPEN`?

andreittr replied:
~~That's true, I think vopen is meant to return opened nodes; the docs are out of date w/ the code, will double-check and update :+1:~~

andreittr replied:
Wait, no, the docs are correct; the `node_vopen` function returns an _unopened_ rootnode, which the `live_vopen` function then passes to `_RSTATE_OPEN`, which in turn calls `nopen` on the rootnode.

mogasergiu commented on lib/ukfs/include/uk/fs/template/node.h:
```diff
@@ -0,0 +1,1977 @@
+				       nod, out_ukfs, &prog);		\
+		if (below && (ret != UKFS_STOP_SYM || n != nod[1]))	\
+			/* n has no live ref & won't be used, forget */	\
+			(node_nforget)(rs->vol->dev, n);		\
```
What is the exact contract for `node_nforget`? Should it be called for every unopened node returned by lookup/vopen that won't be used?

andreittr replied:
Exactly, I mention the 3 possible lifecycles [here](https://github.com/unikraft/unikraft/pull/1729/files#diff-cea516ef7b098d2e0008f4514653b9b4034b9b129799b78e7f5caac083ecd5b8R45).

mogasergiu commented on lib/ukfs/include/uk/fs/template/node.h:
```diff
@@ -0,0 +1,1977 @@
+			ret = drv##_RSTATE_FROMNODE(vs, n, !already_open);\
+			if (unlikely(PTRISERR(ret))) {			\
+				if (already_open)			\
+					(node_nclose)(vs->dev, n);	\
```
Shouldn't it be forgotten (`node_nforget`) otherwise?

andreittr replied:
Indeed, and moreover, since `_RSTATE_OPEN` is tasked with all error handling for the node passed in, I think we're missing one more forget still. Will fix in pass, thanks!

mogasergiu commented on lib/ukfs/include/uk/fs/template/node.h:
```diff
@@ -0,0 +1,1977 @@
+/* Result of opening a volume */					\
+struct UK_FS_TMPL_NODE_OPENDEV(drv) {					\
+	devtype dev; /* Device identifier */				\
+	nodetype rootnode; /* Identifier of root filesystem node */	\
```
Could you clarify in the docs as well what "identifier" means?

If it's just an identifier like an inode number (e.g. `__u64`):
Multiple "copies" of the same nodetype value don't need individual cleanup
`nforget()` is just a hint that "we won't look this up again soon" (cache management)

If it's a counted reference (like a refcounted pointer, a `struct node *`):
Each time `node_fs_lookup` returns a `nodetype`, it increments some internal refcount
`nopen()` and `nforget()` decrement that refcount

andreittr replied:
It indeed means the former; I'll do a pass on the docs to make that clearer.

mogasergiu commented on lib/ukfs/include/uk/fs/template/node.h:
```diff
@@ -0,0 +1,1977 @@
+			}						\
+			n = nod[0];					\
+			cur += prog;					\
+			below = 1;					\
```
Correct me if I am wrong in this case (checking my understanding):
after `node_fs_lookup` returns `nod[0]` (an unopened node ID), and we find it already has an RSTATE (meaning someone already opened that ID) then the existing RSTATE is for the same entity that `node_fs_lookup` just returned. In this case, the ID `nod[0]` is still "in use" by that RSTATE.
We don't need to call `nforget` because the ID won't be freed until that RSTATE is closed.

andreittr replied:
Correct; maybe I should make it clearer in docs that lookup returns _new_ nodes unopened, in-use nodes are returned with unchanged status ("open" in this case).

andreittr commented:
Pushed addressing comments:
- Updated docs for more clarity
- Reworked error handling to have live lookup/create handle node cleanup rather than `_RSTATE_OPEN`

andreittr commented:
Pushed typo fix.

andreittr commented:
Pushed rework of lookup symlink error paths.

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>
Approved-by: Sergiu Moga <sergiu@unikraft.io>
